File: src/twitter_video_dl/update_checker.py
import json
import os
import platform
import time
from pathlib import Path
from typing import Optional, Tuple

import requests

CACHE_EXPIRATION_SECONDS = 86400  # 24 hours


class UpdateChecker:
    """Checks for updates against GitHub Releases API with 24-hour caching."""

    def __init__(self, current_version: str):
        self.current_version = current_version.lstrip("v")
        self.cache_dir = self._get_cache_dir()
        self.cache_file = self.cache_dir / "update_cache.json"

    def _get_cache_dir(self) -> Path:
        """Get platform-specific cache/data folder."""
        home = Path.home()
        sys_platform = platform.system()

        if sys_platform == "Darwin":  # macOS
            return home / "Library" / "Caches" / "twitdl"
        elif sys_platform == "Windows":
            local_appdata = os.getenv("LOCALAPPDATA")
            return (
                Path(local_appdata) / "twitdl"
                if local_appdata
                else home / "AppData" / "Local" / "twitdl"
            )
        else:  # Linux / Unix
            return home / ".cache" / "twitdl"

    def _parse_version(self, version_str: str) -> Tuple[int, ...]:
        """Convert version string to integer tuple for comparison."""
        try:
            return tuple(map(int, version_str.lstrip("v").split(".")))
        except ValueError:
            return (0, 0, 0)

    def check_for_update(self) -> Optional[str]:
        """Check if an update is available.

        Returns:
            The latest version string if an update is available, otherwise None.
        """
        now = time.time()
        cached_version: Optional[str] = None
        last_check: float = 0.0

        # Try to read cached version check info
        if self.cache_file.exists():
            try:
                with open(self.cache_file, "r") as f:
                    data = json.load(f)
                    cached_version = data.get("latest_version")
                    last_check = data.get("last_check", 0.0)
            except Exception:
                pass

        # Check GitHub if cache is expired or missing
        if not cached_version or (now - last_check) > CACHE_EXPIRATION_SECONDS:
            latest_version = self._fetch_latest_version_from_github()
            if latest_version:
                cached_version = latest_version
                try:
                    self.cache_dir.mkdir(parents=True, exist_ok=True)
                    with open(self.cache_file, "w") as f:
                        json.dump(
                            {
                                "latest_version": latest_version,
                                "last_check": now,
                            },
                            f,
                        )
                except Exception:
                    pass

        if cached_version:
            current = self._parse_version(self.current_version)
            latest = self._parse_version(cached_version)
            if latest > current:
                return cached_version

        return None

    def _fetch_latest_version_from_github(self) -> Optional[str]:
        """Fetch latest release tag name from GitHub Releases API."""
        url = (
            "https://api.github.com/repos/reneboygarcia/"
            "twitter_video/releases/latest"
        )
        headers = {"User-Agent": "twitdl-cli"}
        try:
            # Short timeout to avoid blocking the CLI for long if offline
            response = requests.get(url, headers=headers, timeout=1.5)
            if response.status_code == 200:
                data = response.json()
                tag = data.get("tag_name")
                if tag:
                    return tag.lstrip("v")
        except Exception:
            pass
        return None

File: src/update_checker.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};

const CACHE_EXPIRATION_SECONDS: u64 = 86400; // 24 hours

#[derive(Serialize, Deserialize, Debug, Clone)]
struct CacheData {
    latest_version: String,
    last_check: u64,
}

pub struct UpdateChecker {
    current_version: String,
    cache_file: PathBuf,
}

impl UpdateChecker {
    pub fn new(current_version: &str) -> Self {
        let current_version = current_version.trim_start_matches('v').to_string();
        let cache_dir = Self::get_cache_dir();
        let cache_file = cache_dir.join("update_cache.json");
        Self {
            current_version,
            cache_file,
        }
    }

    pub fn new_with_cache_file(current_version: &str, cache_file: PathBuf) -> Self {
        let current_version = current_version.trim_start_matches('v').to_string();
        Self {
            current_version,
            cache_file,
        }
    }

    fn get_cache_dir() -> PathBuf {
        if let Some(mut cache_dir) = dirs::cache_dir() {
            cache_dir.push("twitdl");
            cache_dir
        } else {
            // Fallback to home cache directory
            let mut path = dirs::home_dir().unwrap_or_else(|| PathBuf::from("."));
            path.push(".cache");
            path.push("twitdl");
            path
        }
    }

    fn parse_version(version_str: &str) -> Vec<u32> {
        version_str
            .trim_start_matches('v')
            .split('.')
            .map(|s| s.parse::<u32>().unwrap_or(0))
            .collect()
    }

    pub fn check_for_update(&self) -> Option<String> {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_secs())
            .unwrap_or(0);

        let mut cached_version: Option<String> = None;
        let mut last_check: u64 = 0;

        // Try reading cache
        if self.cache_file.exists() {
            if let Ok(content) = fs::read_to_string(&self.cache_file) {
                if let Ok(cache) = serde_json::from_str::<CacheData>(&content) {
                    cached_version = Some(cache.latest_version);
                    last_check = cache.last_check;
                }
            }
        }

        // Check GitHub if cache is expired or missing
        if cached_version.is_none() || (now - last_check) > CACHE_EXPIRATION_SECONDS {
            if let Some(latest_version) = self.fetch_latest_version_from_github() {
                cached_version = Some(latest_version.clone());
                // Write to cache
                let cache_data = CacheData {
                    latest_version,
                    last_check: now,
                };
                if let Some(parent) = self.cache_file.parent() {
                    let _ = fs::create_dir_all(parent);
                }
                if let Ok(json_str) = serde_json::to_string(&cache_data) {
                    let _ = fs::write(&self.cache_file, json_str);
                }
            }
        }

        if let Some(ref version) = cached_version {
            let current = Self::parse_version(&self.current_version);
            let latest = Self::parse_version(version);
            if latest > current {
                return cached_version;
            }
        }

        None
    }

    fn fetch_latest_version_from_github(&self) -> Option<String> {
        let url = "https://api.github.com/repos/reneboygarcia/twitter_video/releases/latest";

        let agent = ureq::AgentBuilder::new()
            .timeout(std::time::Duration::from_millis(1500))
            .build();

        let response = agent.get(url).set("User-Agent", "twitdl-cli").call();

        if let Ok(res) = response {
            if res.status() == 200 {
                #[derive(Deserialize)]
                struct GithubRelease {
                    tag_name: String,
                }
                if let Ok(release) = res.into_json::<GithubRelease>() {
                    return Some(release.tag_name.trim_start_matches('v').to_string());
                }
            }
        }
        None
    }
}
